File: 19.remove-nth-node-from-end-of-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        vector<ListNode *> vec;
		ListNode *p = head;
		while (p != NULL)
		{
			vec.push_back(p);
			p = p->next;
		}
		ListNode *pl = vec[vec.size() - n];
		if (pl == head)
			return pl->next;
		vec[vec.size() - n - 1]->next = pl->next;
		return head;
    }
};

File: 19.remove-nth-node-from-end-of-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        newHead = ListNode(0)
        newHead.next = head
        self.doRemove(newHead, n)
        return newHead.next

    def doRemove(self, cur: ListNode, n: int) -> int:
        if cur is None:
            return 0
        curN = self.doRemove(cur.next, n)
        if curN == n:
            cur.next = cur.next.next
        return curN + 1
